azure-iot-device/tests/iothub/test_sync_clients.py

```diff
@@ -40,67 +62,154 @@ def test_from_authentication_provider_instantiates_client(
         assert isinstance(client._pipeline, expected_pipeline)
 
+    @pytest.mark.it("Raises ValueError on invalid protocol input")
     @pytest.mark.parametrize("auth_provider", ["SymmetricKey"], ids=[""], indirect=True)
-    def test_from_authentication_provider_bad_input_raises_error_transport_name(
-        self, auth_provider
-    ):
+    # TODO: handle the auth_provider fixture differently so it doesn't result in empty [] in output
+    def test_bad_protocol_input_raises_error(self, auth_provider):
         with pytest.raises(ValueError):
             self.client_class.from_authentication_provider(auth_provider, "bad input")
 
-    def test_instantiation_sets_on_connected_handler_in_transport(self, client):
-        assert client._pipeline.on_transport_connected is not None
-        assert client._pipeline.on_transport_connected == client._on_state_change
 
-    def test_instantiation_sets_on_disconnected_handler_in_transport(self, client):
-        assert client._pipeline.on_transport_disconnected is not None
-        assert client._pipeline.on_transport_disconnected == client._on_state_change
+class SharedClientConnectTests(object):
+    @pytest.mark.it("Begins a 'connect' pipeline operation")
+    def test_calls_pipeline_connect(self, client, pipeline):
+        client.connect()
+        assert pipeline.connect.call_count == 1
 
-    def test_instantiation_sets_on_method_request_received_handler_in_transport(self, client):
-        assert client._pipeline.on_transport_method_request_received is not None
-        assert (
-            client._pipeline.on_transport_method_request_received
-            == client._inbox_manager.route_method_request
-        )
+    @pytest.mark.it("Waits for the completion of the 'connect' pipeline operation before returning")
+    def test_waits_for_pipeline_op_completion(self, mocker, client_manual_cb, pipeline_manual_cb):
+        client = client_manual_cb
+        pipeline = pipeline_manual_cb
+        event_init_mock = mocker.patch.object(threading, "Event")
+        event_mock = event_init_mock.return_value
 
-    def test_state_change_handler_clears_method_request_inboxes_on_disconnect(self, client, mocker):
-        clear_method_request_spy = mocker.spy(client._inbox_manager, "clear_all_method_requests")
-        client._on_state_change("disconnected")
-        assert clear_method_request_spy.call_count == 1
+        def check_callback_completes_event():
+            # Assert waiting for Event to complete
+            assert event_mock.wait.call_count == 1
+            assert event_mock.set.call_count == 0
+
+            # Manually trigger callback
+            cb = pipeline.connect.call_args[1]["callback"]
+            cb()
+
+            # Assert Event is now completed
+            assert event_mock.set.call_count == 1
+
+        event_mock.wait.side_effect = check_callback_completes_event
 
-    def test_connect_calls_transport(self, client, pipeline):
         client.connect()
```

Comment:
>       client.connect() [](start = 1, length = 23)

This function and others with the same pattern seem overly complex.  Reading this code, I can't say if it's correct or not.  All I can say from reading this is: 
1. The client.connect function completes
2. If check_callback_completes_event gets called, then a bunch of things get asserted.  

I understand the way this test is _supposed_ to execute.  I just can't say that it does execute that way without running it through a debugger.  For example, if the implementation of client.connect() was replaced with pass, this test would still pass.  

One easy way to work around this would be to assert that check_callback_completes_event is actually called after client.connect returns.  That sort-of works around the issue.  


Another way, which avoids all the Event patching gymnastics, would be like this:
```
call_complete = False
def t():
    client.connect()
    call_complete=True
threading.Thread(t).start()

assert not call_complete
cb = pipeline.connect.call_args[1]["callback"]
cb()
assert call_complete
```

Plus, I think your event_init_mock is mocking _all_ Event() objects and not just the one you want to mock.

Reply:
The key thing that this test does is show that the callback completes the Event (because Event.set has NOT been called until the callback is triggered). My test runs a side effect when the .wait() is triggered that A) verifies that .wait() has been called (kind of redundant, but makes it more readable) but .set() has NOT been called, B) manually triggers the callback, and C) then shows that after the callback .set() HAS been called. 

This proves the functionality because we know for a fact how Event works, and what .wait() and .set() do. If .wait() has been triggered, we know that it blocks until .set() is also triggered. So we know .wait() triggers (thus blocks by implication), and we know .set() is only triggered as a result of the callback.

Admittedly, as you point out, I'm mocking all Event objects, because there's no way to mock this specific Event that is used in connect since it's created and used only within the function. This would normally be okay, but the limitation is that, yeah, there's no way to prove that the .set() call is happening on the same object that called .wait(). I could add check to the number of calls to the Event constructor though - that would prove that only one Event() object exists, and thus .wait() and .set() are both called on the same instance.

Your test doesn't show that it is ONLY by completion of the callback that the call to connect is able to complete. It's close, but because of the Threading model used, there's no way to be 100% certain that `assert not call_complete` doesn't evaluate to `True` due to timing of the threads - `client.connect()` may just not have finished running yet. I could add a time.sleep(1) in there, but that still wouldn't guarantee it since from a black box perspective, I don't know how long .connect() takes to complete. I don't know that the .connect() call is explicitly blocking on the completion of the callback.

---
In reply to: [291325038](https://github.com/Azure/azure-iot-sdk-python-preview/pull/101#discussion_r291325038) [](ancestors = 291325038)